src/components/FeatureThresholdPanel.tsx

```diff
@@ -81,35 +158,55 @@ export default function FeatureThresholdPanel(inputProps: FeatureThresholdPanelP
     props.onChange(newThresholds);
   };
 
-  /** Handle a threshold getting deleted. */
+  /** Handle a threshold getting deleted via the UI buttons. */
   const onClickedRemove = (index: number): void => {
     const newThresholds = [...props.featureThresholds];
     newThresholds.splice(index, 1);
+    // Delete saved min/max bounds for this feature when it's removed.
+    const thresholdToDelete = props.featureThresholds[index];
+    featureMinMax.current.delete(thresholdToKey(thresholdToDelete));
     props.onChange(newThresholds);
   };
 
-  const selectedFeatures = props.featureThresholds.map((t) => t.featureName);
+  ////// RENDERING ///////////////////
+  // The Select dropdown should ONLY show features that are currently present in the dataset.
   const featureOptions =
-    props.dataset?.featureNames.map((name) => ({ label: props.dataset?.getFeatureNameWithUnits(name), value: name })) ||
-    [];
+    props.dataset?.featureNames.map((name) => ({
+      label: props.dataset?.getFeatureNameWithUnits(name),
+      value: name,
+    })) || [];
+  // Filter out thresholds that no longer match the dataset (feature and/or unit), so we only
+  // show selections that are actually valid.
+  const thresholdsInDataset = props.featureThresholds.filter((t) => {
+    const featureData = props.dataset?.features[t.featureName];
+    return featureData && featureData.units === t.unit;
+  });
+  const selectedFeatures = thresholdsInDataset.map((t) => t.featureName);
 
   const renderListItems = (item: FeatureThreshold, index: number): ReactNode => {
+    // Thresholds are matched on both feature names and units; a threshold must match
+    // both in the current dataset to be valid.
     const featureData = props.dataset?.features[item.featureName];
-    if (!featureData) {
-      return <></>;
-    }
+    const disabled = featureData === undefined || featureData.units !== item.unit;
+    // If the feature is no longer in the dataset, use the saved min/max bounds.
+    const savedMinMax = featureMinMax.current.get(thresholdToKey(item)) || [0, 1];
+    const sliderMin = disabled ? savedMinMax[0] : featureData.min;
+    const sliderMax = disabled ? savedMinMax[1] : featureData!.max;
```

Comment:
should both of these lines use `featuredata!` ?

Reply:
Oops, my mistake! Typescript is smart enough to know that `disabled` checks for the existence of `featureData`, so the `!` isn't needed anymore.